platform/x11/cog-platform-x11.c

```diff
@@ -0,0 +1,896 @@
+
+    wpe_view_backend_dispatch_axis_event (wpe_view_data.backend, &input_event.base);
+#else
+    assert (axis_delta[0] ^ axis_delta[1]);
+
+    struct wpe_input_axis_event input_event = {
+        wpe_input_axis_event_type_motion,
+        event->time,
+        xcb_data.pointer.x, xcb_data.pointer.y,
+        0, 0, 0,
+    };
+
+    if (!!axis_delta[0]) {
+        input_event.axis = 1;
+        input_event.value = axis_delta[0];
+    } else {
+        input_event.axis = 0;
+        input_event.value = axis_delta[1];
+    }
+
+    wpe_view_backend_dispatch_axis_event (wpe_view_data.backend, &input_event);
+#endif
+}
+
+static void
+xcb_handle_button_press (xcb_button_press_event_t *event)
+{
+    int16_t axis_delta[4][2] = {
+        {   0, -20 },
+        {   0,  20 },
+        { -20,   0 },
+        {  20,   0 },
+    };
+
+    switch (event->detail) {
+    case 1:
+    case 2:
+    case 3:
+        xcb_data.pointer.button = event->detail;
+        xcb_data.pointer.state = 1;
+        break;
+    case 4:
+    case 5:
+    case 6:
+    case 7:
+        xcb_handle_axis (event, axis_delta[event->detail - 4]);
+        return;
+    default:
+        return;
+    }
+
+    struct wpe_input_pointer_event input_event = {
+        wpe_input_pointer_event_type_button,
+        event->time,
+        xcb_data.pointer.x, xcb_data.pointer.y,
+        xcb_data.pointer.button,
+        xcb_data.pointer.state
+    };
+
+    wpe_view_backend_dispatch_pointer_event (wpe_view_data.backend, &input_event);
+}
+
+static void
+xcb_handle_button_release (xcb_button_release_event_t *event)
+{
+    switch (event->detail) {
+    case 1:
+    case 2:
+    case 3:
+        xcb_data.pointer.button = event->detail;
+        xcb_data.pointer.state = 0;
+        break;
+    default:
+        return;
+    }
+
+    struct wpe_input_pointer_event input_event = {
+        wpe_input_pointer_event_type_button,
+        event->time,
+        xcb_data.pointer.x, xcb_data.pointer.y,
+        xcb_data.pointer.button,
+        xcb_data.pointer.state
+    };
+
+    wpe_view_backend_dispatch_pointer_event (wpe_view_data.backend, &input_event);
+}
+
+static void
+xcb_handle_motion_event (xcb_motion_notify_event_t *event)
+{
+    xcb_data.pointer.x = event->event_x;
+    xcb_data.pointer.y = event->event_y;
+
+    struct wpe_input_pointer_event input_event = {
+        wpe_input_pointer_event_type_motion,
+        event->time,
+        xcb_data.pointer.x, xcb_data.pointer.y,
+        xcb_data.pointer.button,
+        xcb_data.pointer.state
+    };
+
+    wpe_view_backend_dispatch_pointer_event (wpe_view_data.backend, &input_event);
+}
+
+static void
+on_export_fdo_egl_image(void *data, struct wpe_fdo_egl_exported_image *image)
+{
+    xcb_paint_image (image);
+}
+
+static void
+xcb_process_events (void)
+{
+    xcb_generic_event_t *event = NULL;
+
+    while ((event = xcb_poll_for_event (xcb_data.connection))) {
+        switch (event->response_type & 0x7f) {
```

Comment:
Instead of a magic number it is better use the macros defined by the `xcb/xcb_event.h` header:

```c
/**
 * @brief Bit mask to find event type regardless of event source.
 *
 * Each event in the X11 protocol contains an 8-bit type code.
 * The most-significant bit in this code is set if the event was
 * generated from a SendEvent request. This mask can be used to
 * determine the type of event regardless of how the event was
 * generated. See the X11R6 protocol specification for details.
 */
#define XCB_EVENT_RESPONSE_TYPE_MASK (0x7f)
#define XCB_EVENT_RESPONSE_TYPE(e)   (e->response_type &  XCB_EVENT_RESPONSE_TYPE_MASK)
#define XCB_EVENT_SENT(e)            (e->response_type & ~XCB_EVENT_RESPONSE_TYPE_MASK)
```

In this case we would want:

```suggestion
        switch (XCB_EVENT_RESPONSE_TYPE(event)) {
```

Reply:
In Debian, this header is packaged through the libxcb-util0 package, so that would be a separate dependency. Let's avoid that for now.